### PR TITLE
Threat levels

### LINKED ISSUE #115: Create a GET /threat_levels method
It should return a JSON:
```
[
  {
    "profile_id": 1,
    "threat_levels" :
      [
        {
          "date" : "2016-04-05 13:36:56",
          "scenarios" :
          [
            {
              "similarity": 0.8,
              "threat_level" : 3,
              "scenario_id" : 1
            }
          ]
        }
      ],
    "threat_levels_assessment_runs" :
      [
        {
          "start_date" : "2016-04-05 13:36:56",
          "status" : "running",
          "status_explanation" : "explanation"
        }
      ]
  }
]
```

### PR DESCRIPTION
**Pull request by tbartynski**
Closes #115

### CODE REVIEW

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+    def finished_status_and_explanation
+      {
+        status: :finished,
+        explanation: 'Assessment run finished'
+      }
+    end
+
+    def started?
+      @run.start_date > (Time.now - 1.hour) && @newest_result.nil?
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: What's the purpose of `- 1.hour`?

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    end
+
+    def results
+      @run.threat_assessments.collect{|a| a.results}.flatten
+    end
+
+    def started_status_and_explanation
+      {
+        status: :started,
+        explanation: 'Assessment run started less than hour ago'\
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: ...less than **one** hour ago...

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                 'and did not produce any results'
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: and has not produced any results.

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      }
+    end
+
+    def running?
+      @newest_result ? @newest_result.created_at > (Time.now - 2.hours) : false
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: Why `- 2.hours`?

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    def newest_result_created_between_two_five_hours_ago
+      return false unless @newest_result
+      @newest_result.created_at < (Time.now - 2.hours) &&
+        @newest_result.created_at > (Time.now - 5.hours)
+    end
+
+    def warning_status_and_explanation
+      {
+        status: :warning,
+        explanation: 'Assessment run has not produced result in past two hours'
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: run has not produced any results in the past two hours

File: app/services/threat_level/assessment_run_status_data_builder_service.rb [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+      }
+    end
+
+    def error_status_and_explanation
+      {
+        status: :error,
+        explanation: 'Assessment run has not produced result in past five hours'
```
Comment (nowakowski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: (...) not produced any results in the past five hours

File: app/services/threat_level/assessment_runs_data_builder_service.rb [hunk 7]
```diff
@@ -0,0 +1,18 @@
+module ThreatLevel
+  class AssessmentRunsDataBuilderService
```
Comment (nowakowski) [inline app/services/threat_level/assessment_runs_data_builder_service.rb]: Seems like a very convoluted Java-esque way of doing things. Is this class really necessary? :)

File: app/services/threat_level/calculator_service.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+module ThreatLevel
+  class CalculatorService
+    def self.get(result)
+      return 0 if result.similarity <= 0.3
+      return 1 if result.similarity > 0.3 && result.similarity <= 0.6
```
Comment (nowakowski) [inline app/services/threat_level/calculator_service.rb]: `return 1 if (0.3...0.6) === result.similarity`

File: app/services/threat_level/calculator_service.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
...
+      return 2
```
Comment (nowakowski) [inline app/services/threat_level/calculator_service.rb]: ...or just use a `case` statement

File: spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+        end
+        it 'return status "running" with explanation' do
+          expect(ThreatLevel::AssessmentRunStatusDataBuilderService.get(run)).
+            to eq running_run_status_data
+        end
+      end
+    end
+  end
+
+  context 'run started more hour ago and less than five hours ago' do
```
Comment (nowakowski) [inline spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb]: more than

File: spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    explanation: 'Assessment run finished'
+  }
+end
+
+def started_run_status_data
+  {
+    start_date: run.start_date,
+    status: :started,
+    explanation: 'Assessment run started less than hour ago'\
+                 'and did not produce any results'
```
Comment (nowakowski) [inline spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb]: and has not produced

File: spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+    status: :running,
+    explanation: 'Assessment run has results produced less than two hours ago'
+  }
+end
+
+def warning_run_status_data
+  {
+    start_date: run.start_date,
+    status: :warning,
+    explanation: 'Assessment run has not produced result in past two hours'
```
Comment (nowakowski) [inline spec/services/threat_level/assessment_run_status_data_builder_service_spec.rb]: produced any result in the past two hours

**Comment by nowakowski** [comment]
@tbartynski I believe the cutoffs (1 hour, 2 hours, 5 hours) should not be baked into the controller code, but rather left as configurable parameters in `application.rb` - the same way we've been doing for all other services.

File: app/services/threat_level/assessment_run_status_data_builder_service.rb (hunk 1 above)
Comment (tbartynski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: Status is started if run was started less than 1 hour ago and it has not produced any results yet. I assumed 1 hour is long enough to produce results for threat assessment workflow.

File: app/services/threat_level/assessment_run_status_data_builder_service.rb (hunk 4 above)
Comment (tbartynski) [inline app/services/threat_level/assessment_run_status_data_builder_service.rb]: Comparing is being run once an hour so it should produce some results more less every hour. Of course those 1 hour and 2 hours are arbitrary but they match periods defined in threat assessment workflow in Hyperflow.

File: app/services/threat_level/assessment_runs_data_builder_service.rb (hunk 7 above)
Comment (tbartynski) [inline app/services/threat_level/assessment_runs_data_builder_service.rb]: Actually that is Clean code-sque way of implementing things.

https://www.amazon.com/Clean-Code-Handbook-Software-Craftsmanship/dp/0132350882

If you insist on compressing code into less classes I can do it but don't want to :)

File: app/services/threat_level/calculator_service.rb (hunk 8 above)
Comment (tbartynski) [inline app/services/threat_level/calculator_service.rb]: Applied

**Comment by tbartynski** [comment]
@nowakowski I've implemented most of your remarks. If you find better names for introduces configuration values please do change them and commit.

**Comment by nowakowski** [comment]
@tbartynski Code looks good but changes have occurred in master since you created this PR. Please merge latest master into your branch, run specs and recommit.

**Comment by tbartynski** [comment]
@nowakowski github claims that **This branch has no conflicts with the base branch
Merging can be performed automatically.**
I would trust that we can safely merge my branch int omaster.

**Comment by nowakowski** [comment]
@tbartynski Yes, but you need to make sure all specs are still green after you merge the current master back into your branch. This has nothing to do with git conflicts.

**Comment by tbartynski** [comment]
I did a git rebase master on my branch and every thing is still green.

**Comment by nowakowski** [comment]
Okay, that's good then.

**Comment by nowakowski** [comment]
:+1:

**Comment by nowakowski** [comment]
Deployed on prod.

**Pull request merged**